object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/CalledMethodsPredicateEvaluator.java

```diff
@@ -15,6 +29,118 @@ public CalledMethodsPredicateEvaluator(final Set<String> cmMethods) {
     this.cmMethods = cmMethods;
   }
 
+  /**
+   * Return true iff the boolean formula "lhs -> rhs" is valid, treating all variables as
+   * uninterpreted.
+   *
+   * <p>Used to determine subtyping between two @CalledMethodsPredicate annotations that are not
+   * identical.
+   */
+  public static boolean implies(final String lhsOrig, final String rhsOrig) {
+
+    JavaParser parser = new JavaParser();
+
+    String preamble = "class DUMMY { boolean DUMMY() { return ";
+    String afterward = "; } }";
+
+    String lhs = preamble + lhsOrig + afterward;
+    String rhs = preamble + rhsOrig + afterward;
+
+    CompilationUnit lhsAst = parser.parse(lhs).getResult().orElse(null);
+    CompilationUnit rhsAst = parser.parse(rhs).getResult().orElse(null);
+
+    if (lhsAst == null || rhsAst == null) {
+      return false;
+    }
+
+    Configuration config = Configuration.defaultConfiguration();
+    LogManager log = LogManager.createNullLogManager();
+    ShutdownNotifier notifier = ShutdownManager.create().getNotifier();
+    SolverContext context = null;
+    try {
+      context =
+          SolverContextFactory.createSolverContext(
+              config, log, notifier, SolverContextFactory.Solvers.SMTINTERPOL);
+    } catch (InvalidConfigurationException e) {
+      return false;
+    }
+
+    BooleanFormulaManager booleanFormulaManager =
+        context.getFormulaManager().getBooleanFormulaManager();
+    BooleanFormula lhsBool, rhsBool;
+
+    try {
+      lhsBool = compilationUnitToBooleanFormula(lhsAst, booleanFormulaManager);
+      rhsBool = compilationUnitToBooleanFormula(rhsAst, booleanFormulaManager);
+    } catch (UnsupportedOperationException e) {
+      return false;
+    }
+
+    BooleanFormula satQuery =
+        booleanFormulaManager.not(booleanFormulaManager.implication(lhsBool, rhsBool));
+
+    ProverEnvironment prover =
+        context.newProverEnvironment(SolverContext.ProverOptions.GENERATE_MODELS);
+
+    try {
+      prover.addConstraint(satQuery);
+    } catch (InterruptedException e) {
+      return false;
+    }
+
+    try {
+      return prover.isUnsat();
+    } catch (Exception e) {
+      return false;
+    }
+  }
+
+  /**
+   * Converts a Javaparser AST representing a boolean expression into a boolean formula that could
+   * be passed to an SMT solver.
+   */
+  private static BooleanFormula compilationUnitToBooleanFormula(
+      CompilationUnit ast, BooleanFormulaManager booleanFormulaManager) {
+
+    BlockStmt theBlock = ast.getType(0).getMembers().get(0).asMethodDeclaration().getBody().get();
+
+    com.github.javaparser.ast.expr.Expression theExpression =
+        theBlock.getStatements().get(0).asReturnStmt().getExpression().get();
+
+    try {
+      return expressionToBooleanFormula(theExpression, booleanFormulaManager);
+    } catch (UnsupportedOperationException e) {
+      throw e;
+    }
+  }
+
+  /** Actual implementation of conversion between Javaparser expression and boolean formula. */
```

Comment:
"Actual implementation of" is not very informative.  Maybe the comment would be better off without it.